blssig/src/verifier/mod.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright 2019-2024 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use ahash::{HashMap, HashMapExt};
+use bls_signatures::{PublicKey, Serialize, Signature, verify_messages};
+use filecoin_f3_gpbft::PubKey;
+use filecoin_f3_gpbft::api::Verifier;
+use parking_lot::RwLock;
+use thiserror::Error;
+
+use crate::bdn::BDNAggregation;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum BLSError {
+    #[error("empty public keys provided")]
+    EmptyPublicKeys,
+    #[error("invalid public key length: expected 48 bytes, got {0}")]
+    InvalidPublicKeyLength(usize),
+    #[error("failed to deserialize public key")]
+    PublicKeyDeserialization,
+    #[error("invalid signature length: expected 96 bytes, got {0}")]
+    InvalidSignatureLength(usize),
+    #[error("failed to deserialize signature")]
+    SignatureDeserialization,
+    #[error("BLS signature verification failed")]
+    SignatureVerificationFailed,
+    #[error("mismatched number of public keys and signatures: {pub_keys} != {sigs}")]
+    LengthMismatch { pub_keys: usize, sigs: usize },
+}
+
+/// BLS signature verifier using BDN aggregation scheme
+///
+/// This verifier implements the same scheme used by go-f3/blssig, with:
+/// - BLS12_381 curve
+/// - G1 for public keys, G2 for signatures  
+/// - BDN aggregation for rogue-key attack prevention
+/// - Point cache for performance optimization
+pub struct BLSVerifier {
+    /// Cache for deserialized public key points to avoid expensive repeated operations
+    point_cache: RwLock<HashMap<Vec<u8>, PublicKey>>,
```

Comment:
I recommend using dedicated data structures for caching; here an LRU cache would fit, most likely. Either from https://crates.io/crates/lru or https://crates.io/crates/hashlink (we lately switched to the latter due to lower memory footprint without a visible performance hit)

Reply:
Fixed here: https://github.com/ChainSafe/rust-f3/pull/27/commits/7f9300cbb7bd1885042e84a46b9564345f7a79b5